_pytest/_code/code.py

```diff
@@ -110,7 +110,10 @@ def eval(self, code, **vars):
         """
         f_locals = self.f_locals.copy()
         f_locals.update(vars)
-        return eval(code, self.f_globals, f_locals)
+        try:
```

Comment:
You were absolutely right @RonnyPfannschmidt, try/except around the `eval` was not the proper fix.

I changed the code to deal with the problem explicitly at the place where we try to detect the recursion point. The original cases will be handled exactly like before, except that now instead of blowing up if a comparison fails we display a message indicating the exception raised by the comparison and show the first and last frames of the traceback instead.